hbase-server/src/main/java/org/apache/hadoop/hbase/master/assignment/RegionInTransitionTracker.java

```diff
@@ -0,0 +1,160 @@
+   * other servers.
+   */
+  public void regionCrashed(RegionStateNode regionStateNode) {
+    if (regionStateNode.getRegionInfo().getReplicaId() != RegionInfo.DEFAULT_REPLICA_ID) {
+      return;
+    }
+
+    if (addRegionInTransition(regionStateNode)) {
+      LOG.debug("{} added to RIT list because hosting region server is crashed ",
+        regionStateNode.getRegionInfo().getEncodedName());
+    }
+  }
+
+  /**
+   * Processes a region state change and updates the RIT tracking accordingly. This is the core
+   * method that determines whether a region should be added to or removed from the RIT list based
+   * on its current state and the table's enabled/disabled status. This method should be called
+   * whenever a region state changes get stored to hbase:meta Note: Only default replicas (replica
+   * ID 0) are tracked. Read replicas are ignored.
+   * @param regionStateNode the region state node with the current state information
+   */
+  public void handleRegionStateNodeOperation(RegionStateNode regionStateNode) {
+    // only consider default replica for availability
+    if (regionStateNode.getRegionInfo().getReplicaId() != RegionInfo.DEFAULT_REPLICA_ID) {
+      return;
+    }
+
+    RegionState.State currentState = regionStateNode.getState();
+    boolean tableEnabled = isTableEnabled(regionStateNode.getTable());
+    List<RegionState.State> terminalStates =
+      tableEnabled ? ENABLE_TABLE_REGION_STATE : DISABLE_TABLE_REGION_STATE;
+
+    // if region is merged or split it should not be in RIT list
+    if (
+      currentState == RegionState.State.SPLIT || currentState == RegionState.State.MERGED
+        || regionStateNode.getRegionInfo().isSplit()
+    ) {
+      if (removeRegionInTransition(regionStateNode.getRegionInfo())) {
+        LOG.debug("Removed {} from RIT list as it is split or merged",
+          regionStateNode.getRegionInfo().getEncodedName());
+      }
+    } else if (!terminalStates.contains(currentState)) {
+      if (addRegionInTransition(regionStateNode)) {
+        LOG.debug("{} added to RIT list because it is in-between state, region state : {} ",
+          regionStateNode.getRegionInfo().getEncodedName(), currentState);
+      }
+    } else {
+      if (removeRegionInTransition(regionStateNode.getRegionInfo())) {
+        LOG.debug("Removed {} from RIT list as reached to terminal state {}",
+          regionStateNode.getRegionInfo().getEncodedName(), currentState);
+      }
+    }
+  }
+
+  private boolean isTableEnabled(TableName tableName) {
+    if (tableStateManager != null) {
+      return tableStateManager.isTableState(tableName, TableState.State.ENABLED,
+        TableState.State.ENABLING);
+    }
+    // AssignmentManager calls setTableStateManager once hbase:meta is confirmed online, if it is
+    // still null it means confirmation is still pending. One should not access TableStateManger
+    // till the time.
+    if (TableName.isMetaTableName(tableName)) {
```

Comment:
Maybe here we could add an `assert TableName.isMetaTableName(tableName);`, and then `return true`?

Anyway, throwing a RuntimeException is also acceptable as in production environment we may disable assertion.

Better add something like "CODE-BUG" to tell developers that you may write code wrong.